File: src/main/java/com/example/sk/taskmanagement/service/ReportingService.java
package com.example.sk.taskmanagement.service;

import com.example.sk.taskmanagement.api.CreateCategoryRequest;
import com.example.sk.taskmanagement.domain.*;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public final class ReportingService {

    private final CategoryTaskViewRepository categoryTaskViewRepository;
    private final TaskRepository taskRepository;
    private final CategoryRepository categoryRepository;

    public ReportingService(CategoryTaskViewRepository categoryTaskViewRepository,
                            TaskRepository taskRepository,
                            CategoryRepository categoryRepository) {
        this.categoryTaskViewRepository = categoryTaskViewRepository;
        this.taskRepository = taskRepository;
        this.categoryRepository = categoryRepository;
    }

    public List<CategoryTaskView> getAllCategoryTasksFlat() {
        return categoryTaskViewRepository.findAllCategoryTasks();
    }

    public Category createCategoryWithTasks(CreateCategoryRequest createCategoryRequest) {
        final var createdCategory = categoryRepository.save(new Category(null, createCategoryRequest.name(), Set.of()));

        // create and persist each task
        final var savedTasks = createCategoryRequest.tasks()
                .stream()
                .map(t -> new Task(
                        null,
                        createdCategory.id(),
                        t.description(),
                        Instant.now(),
                        t.dueDate()
                ))
                .map(taskRepository::save)
                .collect(Collectors.toSet());

        // Step 3: Save category again with the tasks included
        final var categoryWithTasks = new Category(createdCategory.id(), createdCategory.name(), savedTasks);
        return categoryRepository.save(categoryWithTasks);
    }
}

File: src/main/kotlin/com/example/sk/taskmanagement/service/ReportingService.kt
package com.example.sk.taskmanagement.service

import com.example.sk.taskmanagement.api.CreateCategoryRequest
import com.example.sk.taskmanagement.api.CreateTaskRequest
import com.example.sk.taskmanagement.domain.Category
import com.example.sk.taskmanagement.domain.CategoryRepository
import com.example.sk.taskmanagement.domain.CategoryTaskView
import com.example.sk.taskmanagement.domain.CategoryTaskViewRepository
import com.example.sk.taskmanagement.domain.Task
import com.example.sk.taskmanagement.domain.TaskRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.Instant

@Service
@Transactional
class ReportingService(
    private val categoryTaskViewRepository: CategoryTaskViewRepository,
    private val taskRepository: TaskRepository,
    private val categoryRepository: CategoryRepository
) {

    fun getAllCategoryTasksFlat(): List<CategoryTaskView> {
        return categoryTaskViewRepository.findAllCategoryTasks()
    }

    fun createCategoryWithTasks(createCategoryRequest: CreateCategoryRequest): Category {
        val createdCategory = categoryRepository.save(Category(
            id = null,
            name = createCategoryRequest.name
        ))

        // create and persist each task
        val savedTasks = createCategoryRequest.tasks.map { taskRequest ->
            Task(
                id = null,
                categoryId = createdCategory.id!!,
                description = taskRequest.description,
                creationDate = Instant.now(),
                dueDate = taskRequest.dueDate
            )
        }.map { taskRepository.save(it) }.toSet()

        // Step 3: Save category again with the tasks included
        val categoryWithTasks = Category(
            id = createdCategory.id,
            name = createdCategory.name,
            tasks = savedTasks
        )
        return categoryRepository.save(categoryWithTasks)
    }
}
